File: prova-c/1021/gabarito.py
from typing import List, Tuple

notes = [10000, 5000, 2000, 1000, 500, 200]
coins = [100, 50, 25, 10, 5, 1]
Distrib = List[Tuple[int, int]]


def money_distrib(cents: int, values: List[int]) -> Distrib:
    distrib = []
    for v in values:
        n = cents // v
        cents -= n * v
        distrib.append((v, n))
    return distrib, cents


def print_money_distrib(distrib: Distrib, name: str) -> None:
    for v, n in distrib:
        print(f"{n} {name} de R$ {v * 0.01:.2f}")


if __name__ == "__main__":
    real, cents = map(int, input().split("."))
    cents = real * 100 + cents

    print("NOTAS:")
    distrib, cents = money_distrib(cents, notes)
    print_money_distrib(distrib, "nota(s)")

    print("MOEDAS:")
    distrib, cents = money_distrib(cents, coins)
    print_money_distrib(distrib, "moeda(s)")

File: prova-c/1021/gabarito.c
#include <stdio.h>

int NOTE_VALUES[6] = {100, 50, 20, 10, 5, 2};
int COIN_VALUES[6] = {100, 50, 25, 10, 5, 1};

int money_distrib(int cents, int *values, int *distrib)
{
    for (int i = 0; i < 6; i++)
    {
        int n = cents / values[i];
        cents -= n * values[i];
        distrib[i] = n;
    }
    return cents;
}

void print_money_distrib(int *distrib, char *name, int *values, float scale)
{
    for (int i = 0; i < 6; i++)
        printf("%d %s de R$ %.2f\n", distrib[i], name, scale * values[i]);
}

void main()
{
    int notes, cents;
    int distrib[6];

    scanf("%d.%d", &notes, &cents);

    printf("NOTAS:\n");
    notes = money_distrib(notes, NOTE_VALUES, distrib);
    print_money_distrib(distrib, "nota(s)", NOTE_VALUES, 1.0);

    printf("MOEDAS:\n");
    money_distrib(cents + 100 * notes, COIN_VALUES, distrib);
    print_money_distrib(distrib, "moeda(s)", COIN_VALUES, 0.01);
}
